refactor: Enforce consistent list[messages] API across agent messaging

priyansh4320 opened the pull request:
<!-- Thank you for your contribution! Please review https://docs.ag2.ai/latest/docs/contributor-guide/contributing before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Supports:
1) conversation API 
https://platform.openai.com/docs/api-reference/conversations/create-items is where OAI support list input.

2) This is also used for response API https://platform.openai.com/docs/api-reference/responses/create#responses_create-conversation

And response API is going to replace chat completion API and all new features will be released through it.

3) also adds support for threads API https://platform.openai.com/docs/api-reference/threads/createThread. 

4) This PR standardizes all agent messaging to a single, predictable list[dict] shape end-to-end, which:
- Reduces branching and simplifies code paths
- Makes hook processing straightforward (hooks always receive normalized dicts)
- Improves support for multimodal content and tool calls
- Provides a cleaner foundation for future messaging enhancements
Changes Made
Type System Updates:
- Updated type hints for send, a_send, receive, a_receive, initiate_chat, a_initiate_chat, generate_reply, and a_generate_reply to accept/return list[dict[str, Any]]
- Updated Agent protocol signatures to match implementation
Runtime Compatibility:
- Added centralized _normalize_message_to_list() for backward compatibility
- Runtime silently accepts legacy str | dict formats and normalizes them
All existing code continues to work without modification
Bug Fixes:
- Fixed infinite loop issue in nested chats
Added test isolation helpers to prevent cross-test contamination from handoff state
Testing:
- Added comprehensive backward compatibility test suites (626 lines sync, 655 lines async)
- Verified compatibility with groupchats, handoffs, tool calling, patterns, groupchat managers, handoff conditions, and nested chats
- Updated experimental hook implementations (reliable.py) to handle normalized format
Backward Compatibility
- Existing code using agent.send({"content": "..."}) continues to work via runtime normalization. Type hints guide new code toward the canonical format while runtime remains flexible.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
closes ag2ai/ag2classic#21 
closes ag2ai/ag2#2124 
<!-- For example: "Closes ag2ai/ag2#1234" -->

## Checks

- [x] I've included any doc changes needed for https://docs.ag2.ai/. See https://docs.ag2.ai/latest/docs/contributor-guide/documentation/ to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

priyansh4320 committed:
feat: consistant API list[messages]

[skip ci] docs: fix outdated docs

fix: recieve,a_recieve for list[messages]

fix: send(),a_send(),_process_message_before_send() handle list[messages]

test: list[messages] on changed function

fix: _process_messages_before_send() handles list[messages]

fix: recieve(), a_recieve() handles list[messages]

fix: send(),a_send() handles list[messages]

fix: pre-commit

fix generate reply for single message

test: generate_reply() with dict[str,Any]

[skip ci] docs: fix outdated docs

Lancetnik reviewed (changes requested):
Great work! However, we should standardize the data and use `messages: list[dict[str, Any]]` instead of `messages: list[dict[str, Any] | dict[str, Any]].` So we need to refactor the current logic to return a `[reply]` instead of `reply` to unify all usage and work with lists consistently.

priyansh4320 commented:
> Great work! However, we should standardize the data and use `messages: list[dict[str, Any]]` instead of `messages: list[dict[str, Any] | dict[str, Any]].` So we need to refactor the current logic to return a `[reply]` instead of `reply` to unify all usage and work with lists consistently.

thanks , will fix this

priyansh4320 committed:
Test/Bugs: add/fix [Testing], bugfixes on Handoffs, nested chats, interoperability

fix: add backward compatibility tests

fix: RandomPattern Handoff Error

fix: tests

fix some tests

fix: tests

fix: conversable agent tests

fix: nested chats

fix:test interoperability

fix: update tool registry with tool signature

fix: tavily tests

test: [sync/async] run , run_group_chat

fix:tests

randombet reviewed (changes requested):
One general comment to avoid introducing the breaking feature here is, we can add list[dict[str, any]] to the union of our public APIs. 
Internally, we could normalize other message format into list[dict[str, any]]. @priyansh4320

randombet commented on autogen/agentchat/agent.py:
```diff
@@ -72,45 +72,49 @@ async def a_send(
     def generate_reply(
         self,
-        messages: list[dict[str, Any]] | None = None,
+        messages: list[dict[str, Any]] | dict[str, Any] | None = None,
```
Why add a dict[str, Any] here? I thought our goal is to migrate the format to list.

priyansh4320 replied:
updated the protocol. thanks for the review

priyansh4320 committed:
fix: Agent Protocal

priyansh4320 committed:
fix: Documentation

randombet commented on autogen/agentchat/agent.py:
```diff
@@ -40,14 +40,14 @@ def description(self) -> str:
     def send(
         self,
-        message: dict[str, Any] | str,
+        message: list[dict[str, Any]] | str,
```
change to """list[dict[str, Any]] | dict[str, Any] | str""" to avoid the breaking change

priyansh4320 replied:
thank you for the review, I was asked to keep the current change, the changes you have asked "list[dict[str, Any]] | dict[str, Any] | str", had been made in the past. and the current change is updating that change.
therefore, I am resolving this Conversation.

priyansh4320 commented:
@Lancetnik drafting this, made a mistake while rebasing and want to test my PR once more , based on claude review.

priyansh4320 committed:
fix: normalize msg to list[messages]

priyansh4320 committed:
fix: test

priyansh4320 committed:
test: backward compatibility

priyansh4320 committed:
fix: _validator_structured_output_hook for mormaization of messages

priyansh4320 committed:
Documentation: update for list[messages]

this PR aims to refactor the API consistancy to be enforced to list[messages] i.e, type-hints : list[dict[str,Any]]
- I have updated send , a_send , recieve , a_recieve, process_message_before_send with consistant API across .
- updated protocals in agents.py to match list[messages]
- add async and sync backward compatibility tests and skipped them intentionally since integration tests.
- handled groupchats
- updated validated_structured_output_hook kept messages as dict[str,Any] intentionally, so ignore this type hint
- updated documentation docstrings.

priyansh4320 committed:
fix: add checks for processed message and fix tests

@claude
- fixed the tests
- conditional skipped the backward compatibility tests
- avoided overloads intentional
- dismissed: (as not needed type check works well) The protocol signatures declare message: list[dict[str, Any]], but the implementation accepts str | dict | list. Type checkers will flag this mismatch.

priyansh4320 committed:
fix: test creds

priyansh4320 committed:
fix: breaking API contract for list[messages]

this commit fixes the breakage in API contract with strict typing with overloads.
while for single message as well as str message types in processing is handled with normalized message method.

priyansh4320 committed:
prevent: breaking API contract

this commit fixes the breakage in API contract with strict typing with overloads.

while for single message as well as str message types in processing is handled with normalized message method where it is where the message is made consistent for list[message] therefore now avoiding the breaking change with contract.

priyansh4320 committed:
fix: return types and hook validation

priyansh4320 committed:
fix: handle custom fields in normalize message to oai

priyansh4320 committed:
fix: filter NoneTypes

priyansh4320 committed:
fix: add docstring

priyansh4320 committed:
test: update list[message]

priyansh4320 committed:
fix: typos

let's merge this

priyansh4320 committed:
Merge branch 'main' into update-generate-reply-function

Lancetnik reviewed (approved):
LGTM
Thank you for your work!

randombet commented on autogen/agentchat/agent.py:
```diff
@@ -38,66 +38,139 @@ def description(self) -> str:
+
+    def send(
+        self,
+        message: list[dict[str, Any]] | str | dict[str, Any],
```
It overlaps with line 44. Why do we need two methods?

randombet commented on autogen/agentchat/agent.py:
```diff
@@ -38,66 +38,139 @@ def description(self) -> str:
 
+    async def a_send(
+        self,
+        message: list[dict[str, Any]] | str | dict[str, Any],
```
Same here. You only need this method as this is the super set of line 83 and line 75

Lancetnik replied:
Moreover `overload` has no sense if return result the same. We can just use the original signature

randombet commented on autogen/agentchat/conversable_agent.py:
```diff
@@ -1023,6 +1023,29 @@ def _message_to_dict(message: dict[str, Any] | str) -> dict:
             return dict(message)
 
+    @staticmethod
+    def _normalize_message_to_list(
```
This method should go to utils.py or a new file under agentchat/ 
conversable_agent.py is already bloated.

And add unit tests for this method is a must as this is critical method.

priyansh4320 committed:
Merge branch 'main' into update-generate-reply-function

randombet commented on test/agentchat/test_nested.py:
```diff
@@ -331,7 +332,7 @@ def is_termination(msg) -> bool:
     assistant = autogen.AssistantAgent(
         "Assistant_In_Group_1",
     )
-    MockAgentReplies(["Assistant_In_Group_1 message 1"]).add_to_agent(assistant)
```
Same here

priyansh4320 committed:
fix: Remove overloads and move helper methods to utils

priyansh4320 committed:
test: add helper methods tests

priyansh4320 commented on autogen/agentchat/conversable_agent.py:
```diff
@@ -1066,35 +1089,68 @@ def _append_oai_message(
-            )
-        return message
+
+        if isinstance(message, list):
```
thanks for noticing, removed the check

priyansh4320 committed:
fix: add handle list[messages] in _process_recieved_messages

priyansh4320 commented:
> 1. Why do we need test/test_backward_compatibility_sync.py and test/test_backward_compatibility_sync.py? We need to test the list[message] cases, and backward compatibility should be covered from previous tests. We only add the missing part if not exists.

- both files are robust testing covering each pattern and covering various integrations from 2 agents to groupchats and tool calls as well.
benifits:
- better bug resolution , as well as helps in robust refactoring.

priyansh4320 committed:
test: send test with all types list[message],str,dict

priyansh4320 commented:
> 2\. Remove website/docs/contributor-guide/Migration-Guide.mdx as this is not public doc. Add instruction in user-guide

this change was suggested by claude , updated the changes

priyansh4320 committed:
docs: updated migration.mdx changes to -> updates.mdx

priyansh4320 committed:
fix: update groupchat for delta messages

randombet commented on autogen/agentchat/agent.py:
```diff
@@ -38,66 +38,139 @@ def description(self) -> str:
+    @overload
+    def send(
+        self,
+        message: list[dict[str, Any]],
```
Just a single send method with 
message: list[dict[str, Any]] | dict[str, Any] | str

priyansh4320 committed:
docs: update hook docs

priyansh4320 committed:
docs: update

priyansh4320 committed:
fix: tavily test

priyansh4320 commented on test/tools/experimental/tavily/test_tavily.py:
```diff
@@ -40,9 +41,12 @@ def test_initialization(self, use_internal_auth: bool) -> None:
-            with pytest.raises(ValueError) as exc_info:
-                TavilySearchTool(tavily_api_key=None)
-            assert "tavily_api_key must be provided" in str(exc_info.value)
+            with patch.dict(os.environ, {}, clear=False):
```
removed the patch

priyansh4320 committed:
fix: cover conversable agent tests

priyansh4320 committed:
test: remove redundant test

priyansh4320 commented:
Addressed all the comments and changes:
- updated group chats to handle delta messages  and added test for these in test_groupchats.py
- all tests are passing 
- Added documentation
- removed redundant test, and overloads
- moved helper methods from conversable_agents.py to utils.py and added test in test_agentchat_utils.py

@randombet can you please test this PR and approve if everything is okay?

randombet reviewed (changes requested):
Please add integration tests (not only unit test) for two agent chat, Nested Chat
, group chat, group chat with Orchestration Patterns.
Cover ConversableAgent::run, initiate_chat, initiate_chats, initiate_group_chat, run_group_chat and their async counterparts.

priyansh4320 committed:
test: integrations [conversable_Agent, groupchat, nested , tool calls]

priyansh4320 committed:
test: async integration

priyansh4320 committed:
test: conversable_Agent

priyansh4320 committed:
fix: cleanup

priyansh4320 committed:
test: integration

priyansh4320 committed:
Merge branch 'main' into update-generate-reply-function

priyansh4320 commented:
@randombet I have cross-checked, my integration tests are correct, where real configuration is being used with credentials_gpt_4o_mini.llm_config, credentials_gpt_4o.llm_config,
- All the integration tests are marked with @pytest.mark.integration :

Here is a list of all the integration tests I wrote; the rest are unit tests:
List of integration tests
test_groupchat.py 
Synchronous Tests:
1. test_group_chat_manager_with_multiple_agents_integration
2. test_group_chat_round_robin_pattern_integration
3. test_group_chat_auto_pattern_integration
4. test_run_group_chat_round_robin_integration
5. test_run_group_chat_auto_pattern_integration
Asynchronous Tests:
1. test_group_chat_manager_with_multiple_agents_async_integration
2. test_group_chat_round_robin_pattern_async_integration
3. test_group_chat_auto_pattern_async_integration
4. test_a_run_group_chat_round_robin_async_integration
5. test_a_run_group_chat_auto_pattern_async_integration
6. test_group_chat_random_pattern_async_integration

test_conversable_agent.py 
Synchronous Tests:
1. test_single_agent_generate_reply_integration
2. test_two_agent_chat_with_calculator_tool_integration
3. test_agent_run_method_single_integration
4. test_agent_run_method_two_agents_integration
Asynchronous Tests:
1. test_single_agent_generate_reply_async_integration
2. test_two_agent_chat_with_calculator_tool_async_integration
3. test_agent_a_run_method_single_async_integration
4. test_agent_a_run_method_two_agents_async_integration

test_chats.py 
Multi-Turn Conversation Tests:
1. test_complex_multi_turn_conversation_integration
2. test_complex_multi_turn_conversation_async_integration

test_tool_calls.py 
Tool and Math Integration Tests:
1. test_eval_math_responses
2. test_eval_math_responses_api_style_function

Group Chat with Tools Tests:
1. test_handoffs_and_tools_integration
2. test_mixed_tools_scenario_integration


Advanced Nested Chat Integration Tests:
1. test_nested
2. test_conditional_handoffs_with_llm_conditions_integration
3. test_nested_chat_target_integration
4. test_terminate_target_functionality_integration

Async Advanced Nested Chat Integration Tests:
1. test_conditional_handoffs_with_llm_conditions_async_integration
2. test_nested_chat_target_async_integration
3. test_terminate_target_functionality_async_integration

randombet commented:
@priyansh4320 
Could you separate the tests into separate files? Any code over 500 lines is bloated and we should avoid make it worse.
Also separate unit tests and integration tests in one file.

In the integration tests for your case, please use at lease 2 messages in a list as input and verify with meaningful assert
e.g. the following statements can't justify testing purpose.
`
assert result is not None
assert last_agent is not None
`

randombet commented on autogen/agentchat/conversable_agent.py:
```diff
@@ -1023,6 +1029,29 @@ def _message_to_dict(message: dict[str, Any] | str) -> dict:
         else:
             return dict(message)
 
+    # @staticmethod
```
remove this part

randombet commented on autogen/agentchat/groupchat.py:
```diff
@@ -1308,7 +1319,11 @@ def run_chat(
+
+            # NEW: Get the delta messages for the next iteration
+            # (this will be a list with typically one message - the reply)
+            delta_messages = [self.last_message(speaker)]
```
This is incorrect. last_message returns the last one message instead of delta, which could be more than one message.

randombet reviewed (changes requested):
The way to handle last delta messages are incorrect.
@priyansh4320 You need to think through how to get the last delta messages. One way is to add new variable in conversable agent to store the # of processed_messages.

Also notice current GeminiClient is stateful and only send last one messages now. You also need to make sure GeminiClient will send last delta messages.

priyansh4320 committed:
Merge branch 'main' into update-generate-reply-function

marklysze commented:
@priyansh4320 I'll close this as we move to V1.0 with beta.
